File: backend/krushr/internal/handlers/authentication.go
package handlers

import (
	"net/http"

	"github.com/stanhoenson/krushr/internal/constants"
	"github.com/stanhoenson/krushr/internal/env"
	"github.com/stanhoenson/krushr/internal/models"
	"github.com/stanhoenson/krushr/internal/services"
	"github.com/stanhoenson/krushr/internal/wrappers"
	"github.com/gin-gonic/gin"
)

func signUp(c *gin.Context) {
	var signUpBody models.SignUpBody

	if err := c.BindJSON(&signUpBody); err != nil {
		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	createdUser, err := services.CreateUserFromSignUpBody(&signUpBody)
	if err != nil {
		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": "Error signing up"})
		return
	}

	c.IndentedJSON(http.StatusOK, createdUser)
}

func signIn(c *gin.Context) {
	var signInBody models.SignInBody

	if err := c.BindJSON(&signInBody); err != nil {
		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	jwt, err := services.Authenticate(&signInBody)
	if err != nil {
		c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Error signing in"})
		return
	}

	c.SetCookie("jwt", jwt, int(constants.TokenValidityPeriod.Seconds()), "/", env.Domain, true, true)

	c.Status(http.StatusOK)
}

func signOut(c *gin.Context) {
	c.SetCookie("jwt", "", 0, "/", env.Domain, true, true)

	c.Status(http.StatusOK)
}

func RegisterAuthenticationRoutes(r *gin.Engine) {
	routes := r.Group("/authentication")
	{
		routes.POST("/sign-in", signIn)
		routes.GET("/sign-out", signOut)
		routes.POST("/sign-up", wrappers.RoleWrapper([]string{constants.AdminRoleName}, signUp))
	}
}

File: frontend/krushr/src/requests/authentication.ts
import axios, { AxiosError } from "axios";
import type { User } from "../types/models";
import type { SignInBody, SignUpBody } from "../types/request-bodies";
import { BASE_URL } from "./endpoints";

export async function signUp(signUpBody: SignUpBody): Promise<User> {
  const response = await axios.post<User>(
    `${BASE_URL}/authentication/sign-up`,
    signUpBody,
    { withCredentials: true }
  );
  return response.data;
}

export async function signIn(signInBody: SignInBody): Promise<string> {
  const response = await axios.post<string>(
    `${BASE_URL}/authentication/sign-in`,
    signInBody,
    { withCredentials: true }
  );
  return response.data;
}

export async function signOut(): Promise<any> {
  const response = await axios.get(`${BASE_URL}/authentication/sign-out`, {
    withCredentials: true,
  });
  return response.data;
}
